pylabrobot/resources/functional.py

```diff
@@ -0,0 +1,59 @@
+import os
+import json
+import random
+from collections import deque
+from typing import List, Optional
+from pylabrobot.resources.tip_rack import TipRack
+
+
+async def linear_tip_spot_generator(tip_racks: List[TipRack],
+                                    cache_file_path: Optional[str]=None):
+  """ Tip spot generator with disk caching. Linearly iterate through all tip spots and
+  raise StopIteration when all spots have been used. """
+  tip_rack_idx, tip_spot_idx = 0, 0
+  if cache_file_path is not None and os.path.exists(cache_file_path):
+    with open(cache_file_path, "r", encoding="utf-8") as f:
+      data = json.load(f)
+      tip_rack_idx, tip_spot_idx = data["tri"], data["tsi"]
+      print("loaded tip idx from disk:", data)
+
+  while tip_rack_idx < len(tip_racks):
+    while tip_spot_idx < 96:
+      tip_spot_idx += 1
+      if cache_file_path is not None:
+        with open(cache_file_path, "w", encoding="utf-8") as f:
+          json.dump({"tri": tip_rack_idx, "tsi": tip_spot_idx}, f)
```

Comment:
why not just spell out `tip_rack_idx` and `tip_spot_idx`? abbreviations are the enemy of readability

Reply:
resource style code artifacts 😅